Lib/asyncio/proactor_events.py

```diff
@@ -227,18 +220,30 @@ def _loop_reading__on_eof(self):
         if not keep_open:
             self.close()
 
-    def _loop_reading(self, fut=None):
-        self._loop_reading_cb(fut)
-
-    def _loop_reading__data_received(self, fut):
-        if (fut is not None and
-                getattr(fut, '__asyncio_cancelled_on_pause__', False)):
+    def _data_received(self, data):
+        if self._paused:
+            # Don't call any protocol method while reading is paused.
+            # The protocol will be called on resume_reading().
+            assert self._pending_data is None
+            self._pending_data = data
             return
 
-        if self._paused:
-            self._reschedule_on_resume = True
+        if not data:
+            self._eof_received()
             return
 
+        if isinstance(self._protocol, protocols.BufferedProtocol):
+            try:
+                sslproto._feed_data_to_bufferred_proto(self._protocol, data)
```

Comment:
I'd also move `_feed_data_to_bufferred_proto` helper to `asyncio/protocols.py` so that this is less confusing for whoever reads the code.

Reply:
Right, done.